posts/2024-11-01-gsoc-2024-results.md

```diff
@@ -0,0 +1,119 @@
+
+> He designed and built one of our most-requested features, and produced design prototypes of several more features our users would love. He also observed that writing quality CLI and functional tests was hard, so he overhauled our test system to make better tests easier to make. Future work on cargo-semver-checks will be much easier thanks to the work Max put in this summer.
+
+Great work, Max!
+
+### Implementation of a Faster Register Allocator For Cranelift
+
+- Contributor: [Demilade Sonuga](https://github.com/d-sonuga)
+- Mentors: [Chris Fallin](https://github.com/cfallin) and [Amanieu d'Antras](https://github.com/Amanieu)
+- [Final report](https://d-sonuga.netlify.app/gsoc/regalloc-iii/)
+
+TODO
+
+### Improve Rust benchmark suite
+
+- Contributor: [Eitaro Kubotera](https://github.com/s7tya)
+- Mentor: [Jakub Beránek](https://github.com/kobzol)
+- [Final report](https://gist.github.com/s7tya/96dc1cae4ca9e59841a95f0f48d023d6)
+
+This project was relatively loosely defined, with the overarching goal of improving the user interface of the [Rust compiler benchmark suite](https://github.com/rust-lang/rustc-perf). Eitaro tackled this challenge from various angles at once. He improved the visualization of runtime benchmarks, which were previously a second-class citizen in the benchmark suite, by adding them to our [dashboard](https://perf.rust-lang.org/dashboard.html) and by implementing [historical charts](https://github.com/rust-lang/rustc-perf/pull/1922) of runtime benchmark results, which help us figure out how is a given benchmark behaving over a longer time span.
+
+Another improvement that he has worked on was embedding a profiler trace visualizer directly within the `rustc-perf` website. This was a challenging task, which required him to evaluate several visualizers and figure out a way how to include them within the source code of the benchmark suite in a non-disruptive way. In the end, he managed to integrate [Perfetto](https://ui.perfetto.dev/) within the suite website, and also performed various [optimizations](https://github.com/rust-lang/rustc-perf/pull/1968) to improve the performance of loading compilation profiles.
+
+Last, but not least, Eitaro also created a completely new user interface for the benchmark suite, which runs entirely in the [terminal](https://github.com/rust-lang/rustc-perf/pull/1955). Using this interface, Rust compiler contributors can examine the performance of the compiler without having to start the rustc-perf website, which can be challenging to deploy locally.
+
+Apart from the mentioned contributions, Eitaro also made a lot of other smaller improvements to various parts of the benchmark suite. Thank you for all your work!
+
+### Move cargo shell completions to Rust
+
+- Contributor: [shanmu](https://github.com/shannmu)
+- Mentor: [Ed Page](https://github.com/epage)
+- [Final report](https://hackmd.io/@PthRWaPvSmS_2Yu_GLbGpg/Hk-ficKpC)
+
+TODO
```

Comment:
Cargo's completion scripts have been hand maintained and frequently broken when changed.  The goal for this effort was to have the completions automatically generated from the definition of Cargo's command-line with extension points for the dynamically generated results.

shanmu took the prototype for clap's dynamic completions and got it working and tested for common shells as well as extended the parser to cover more cases.  They then added extension points for CLI's to provide custom completion results.

In the next phase, shanmu added this to nightly Cargo and added different custom completers to match what the hand written completions do.

See the [instructions](https://doc.rust-lang.org/nightly/cargo/reference/unstable.html#native-completions) for trying this out!  That also lists where you can provide feedback.

For whats left before this can be stabilized, see
- [clap#3166](https://github.com/clap-rs/clap/issues/3166)
- [cargo#14520](https://github.com/rust-lang/cargo/issues/14520)

Reply:
Should we link individuals to their GitHub handles?

Reply:
Thanks a lot! I did some brief editing on this and also added and example, so that people see how cool this is :sunglasses: Let me know what do you think. https://github.com/rust-lang/blog.rust-lang.org/pull/1419/commits/468602997358a5c8e90e49fc16214399d1689d03

Reply:
> Should we link individuals to their GitHub handles?

I added a link to shanmu (plus some tool links), but I don't think that we need to repeat the links everywhere.

Reply:
Yeah, I left the link off because it was already in the "Contributor" line but don't mind if it also appears in the body.